File: backend/app/schemas/contracts.py
from typing import List, Literal, Optional
from pydantic import BaseModel, Field

# Common Literals
CompetencyLevel = Literal["none", "basic", "intermediate", "advanced"]
RequiredLevel = Literal["basic", "intermediate", "advanced"]
DomainType = Literal["statistical", "technical", "digital_governance", "behavioural"]
GapSeverity = Literal["low", "medium", "high"]

# --- Profile Endpoints ---
class ProfileRequest(BaseModel):
    designation: str = Field(..., json_schema_extra={"example": "Statistical Investigator"})
    department: str = Field(..., json_schema_extra={"example": "National Sample Survey Office"})
    job_role: str = Field(..., json_schema_extra={"example": "Statistical Investigator"})
    experience_years: float = Field(..., json_schema_extra={"example": 4})
    education: str = Field(..., json_schema_extra={"example": "M.Sc. Statistics"})
    prior_trainings: List[str] = Field(default_factory=list, json_schema_extra={"example": ["Basic Python", "Survey Sampling"]})

class CompetencyItem(BaseModel):
    node_id: str
    current_level: CompetencyLevel

class ProfileResponse(BaseModel):
    profile_id: str
    competencies: List[CompetencyItem]


# --- Roadmap Endpoints ---
class RoadmapRequest(BaseModel):
    profile_id: str = Field(..., json_schema_extra={"example": "prof_12345"})
    job_role: str = Field(..., json_schema_extra={"example": "Statistical Investigator"})

class MatchedCourse(BaseModel):
    course_id: str
    title: str
    duration_hours: float

class RoadmapNode(BaseModel):
    node_id: str
    name: str
    domain: DomainType
    current_level: CompetencyLevel
    required_level: RequiredLevel
    gap_severity: GapSeverity
    matched_courses: List[MatchedCourse]

class RoadmapResponse(BaseModel):
    roadmap: List[RoadmapNode]


# --- Quiz Endpoints ---
class QuizRequest(BaseModel):
    content_text: Optional[str] = Field(default=None, json_schema_extra={"example": "Sampling error decreases as sample size increases in simple random sampling."})
    node_id: Optional[str] = Field(default=None, json_schema_extra={"example": "stat-sampling-101"})

class QuizQuestion(BaseModel):
    node_id: str = Field(default="stat-sampling-101", json_schema_extra={"example": "stat-sampling-101"})
    subtopic: Optional[str] = Field(default=None, json_schema_extra={"example": "Simple random sampling"})
    question: str
    options: List[str]
    correct_index: int
    explanation: str

class QuizResponse(BaseModel):
    questions: List[QuizQuestion]
    node_id: Optional[str] = None
    subtopics_tested: List[str] = Field(default_factory=list)

# --- Quiz Submission & Mastery Endpoints ---
class QuizAnswerItem(BaseModel):
    node_id: str
    is_correct: bool
    subtopic: Optional[str] = None
    user_answer: Optional[str] = None

class QuizSubmitRequest(BaseModel):
    profile_id: str = Field(..., json_schema_extra={"example": "prof_12345"})
    answers: List[QuizAnswerItem]
    node_id: Optional[str] = None

class MasteryUpdateItem(BaseModel):
    node_id: str
    mastery: float
    current_level: CompetencyLevel
    last_reviewed: str

class QuizSubmitResponse(BaseModel):
    profile_id: str
    mastery_updates: List[MasteryUpdateItem]
    objective_coverage_pct: Optional[float] = None
    covered_subtopics: List[str] = Field(default_factory=list)
    missed_subtopics: List[str] = Field(default_factory=list)
    feedback: Optional[str] = None


# --- Employee Dashboard Endpoints ---
class CompetencySummaryItem(BaseModel):
    node_id: str
    name: str
    mastery: float
    last_reviewed: str

class RevisionSuggestion(BaseModel):
    node_id: str
    name: str
    reason: str

class EmployeeDashboardResponse(BaseModel):
    profile_id: str
    competency_summary: List[CompetencySummaryItem]
    learning_hours_logged: float
    overall_progress_pct: float
    revision_suggestions: List[RevisionSuggestion]


# --- Dashboard Admin Endpoints ---
class AdminEmployeeItem(BaseModel):
    profile_id: str
    name: str
    department: str
    avg_gap_severity: GapSeverity
    top_gaps: List[str]

class AdminDashboardResponse(BaseModel):
    employees: List[AdminEmployeeItem]


# --- Knowledge Hub Endpoints ---
class KnowledgeDocument(BaseModel):
    document_id: str
    filename: str
    file_type: str
    size_bytes: int
    uploaded_at: str
    status: str = "ready"

class KnowledgeDocumentListResponse(BaseModel):
    documents: List[KnowledgeDocument]


# --- Practice Microdata Endpoints ---
class PracticeDatasetResponse(BaseModel):
    dataset_name: str
    data_type: str
    source_note: str
    survey_name: str
    linked_competency_node: Optional[str] = None
    target_job_role: Optional[str] = None
    total_records: int
    quality_issue_count: int
    records: List[dict]

File: frontend/lib/types/contracts.ts
export type CompetencyLevel = "none" | "basic" | "intermediate" | "advanced";
export type RequiredLevel = "basic" | "intermediate" | "advanced";
export type DomainType = "statistical" | "technical" | "digital_governance" | "behavioural";
export type GapSeverity = "low" | "medium" | "high";

export type ProfileRequest = {
  designation: string;
  department: string;
  job_role: string;
  experience_years: number;
  education: string;
  prior_trainings: string[];
};

export type Competency = {
  node_id: string;
  current_level: CompetencyLevel;
};

export type ProfileResponse = {
  profile_id: string;
  competencies: Competency[];
};

export type RoadmapRequest = {
  profile_id: string;
  job_role: string;
};

export type Course = {
  course_id: string;
  title: string;
  duration_hours: number;
};

export type RoadmapItem = {
  node_id: string;
  name: string;
  domain: DomainType;
  current_level: CompetencyLevel;
  required_level: RequiredLevel;
  gap_severity: GapSeverity;
  matched_courses: Course[];
};

export type RoadmapResponse = {
  roadmap: RoadmapItem[];
};

export type QuizQuestion = {
  node_id?: string;
  subtopic?: string;
  question: string;
  options: string[];
  correct_index: number;
  explanation: string;
};

export type QuizResponse = {
  questions: QuizQuestion[];
  node_id?: string;
  subtopics_tested?: string[];
};

export type QuizAnswerItem = {
  node_id: string;
  is_correct: boolean;
  subtopic?: string;
  user_answer?: string;
};

export type QuizSubmitRequest = {
  profile_id: string;
  answers: QuizAnswerItem[];
  node_id?: string;
};

export type MasteryUpdateItem = {
  node_id: string;
  mastery: number;
  current_level: CompetencyLevel;
  last_reviewed: string;
};

export type QuizSubmitResponse = {
  profile_id: string;
  mastery_updates: MasteryUpdateItem[];
  objective_coverage_pct?: number;
  covered_subtopics?: string[];
  missed_subtopics?: string[];
  feedback?: string;
};

export type CompetencySummaryItem = {
  node_id: string;
  name: string;
  mastery: number;
  last_reviewed: string;
};

export type RevisionSuggestion = {
  node_id: string;
  name: string;
  reason: string;
};

export type CourseProgressItem = {
  course_id: string;
  title: string;
  code?: string;
  provider?: string;
  domain?: string;
  competency?: string;
  difficulty?: string;
  duration?: string;
  progress_pct: number;
  completed_modules: number;
  total_modules: number;
  rating?: string;
  reason?: string;
};

export type EmployeeDashboard = {
  profile_id: string;
  competency_summary: CompetencySummaryItem[];
  learning_hours_logged: number;
  overall_progress_pct: number;
  readiness_score?: number;
  target_role_coverage?: number;
  revision_suggestions: RevisionSuggestion[];
  active_courses?: CourseProgressItem[];
};

export type AdminEmployee = {
  profile_id: string;
  name: string;
  department: string;
  avg_gap_severity: GapSeverity;
  top_gaps: string[];
};

export type AdminOrganizationalGapItem = {
  node_id: string;
  name: string;
  gap_severity: GapSeverity;
  progress: number;
};

export type AdminDashboard = {
  employees: AdminEmployee[];
  organizational_gaps?: AdminOrganizationalGapItem[];
};

export type KnowledgeDocument = {
  document_id: string;
  filename: string;
  file_type: string;
  size_bytes: number;
  uploaded_at: string;
  status: string;
};

export type KnowledgeDocumentListResponse = {
  documents: KnowledgeDocument[];
};

export type PracticeDatasetResponse = {
  dataset_name: string;
  data_type: string;
  source_note: string;
  survey_name: string;
  linked_competency_node?: string;
  target_job_role?: string;
  total_records: number;
  quality_issue_count: number;
  records: Record<string, any>[];
};

export interface EmployeeRecord {
  profile_id: string;
  name: string;
  designation: string;
  department: string;
  job_role: string;
  experience_years: number;
  education: string;
  prior_trainings: string[];
  initials?: string;
  cadreCode?: string;
  employeeCode?: string;
  email?: string;
  location?: string;
}



